### ISSUE TITLE
Problem with third-party in Xcode 10 building to iOS physical device

### ISSUE THREAD

**Issue by dlimx** [facebook/react-native#19839]
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment:
```
Environment:
  OS:  macOS 10.14
  Node:  8.11.2
  Yarn:  1.7.0
  npm:  5.6.0
  Watchman:  4.7.0
  Xcode:  Xcode 10.0 Build version 10L176w
  Android Studio:  2.3 AI-162.4069837

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: 0.55.4 => 0.55.4
```
__
## Description
`third-party` encounters a semantic issue during Xcode builds to physical iOS device. The error is `No member named '__rip' in '__darwin_arm_thread_state64'`

## Reproducible Demo

On macOS Mojave with Xcode 10.
```
react-native init Test
cd Test
yarn start
Xcode build to device
```

Fails due to `No member named '__rip' in '__darwin_arm_thread_state64'`

**Comment by hramos** [comment]
Can you try using the latest 0.56 RC? That release contains some related fixes and we'd like to ensure the issue you're describing is not something that is already fixed.

**Comment by dlimx** [comment]
I've applied manual fixes for `libfishook.a` and `glog-0.3.4` - it builds fine to non-device simulators. I'll definitely confirm tomorrow though if it's something that's already been fixed.

**Comment by jumarmartin** [comment]
Not @dlimx, but can report the same. Using the latest RC.
```
Current Environment: 
macOS: 10.14 build 18A314h
node: 9.4.0
yarn: 1.7.0
npm: 6.1.0
watchman: 4.9.0
Xcode: 10.0 build 10L177m
Android Studio: Build #AI-173.4670197

react: 16.3.2
react-native-cli: 2.0.1
react-native: 0.56.0-rc.2
```
This error also occurs when building for a "Generic iOS Device" 
The error seem to stem from `signalhandler.cc` in the `glog` folder, on line 78 
`return (void*)context->PC_FROM_UCONTEXT;`

I saw that there was a version bump to rc.3, I couldn't use `react-native-git-upgrade` nor `react-native upgrade` to upgrade to that version.

**Comment by tanabs18** [comment]
@hramos I'm having the same issue on rc.2, rc.3, and rc.4

**Comment by PrimeObjects** [comment]
same issue here. Any solution?

**Comment by JoseVf** [comment]
It happens to me on a recently created RN project, but I'm using xCode 9.4
IDEs:
      `Xcode: 9.4.1/9F2000 - /usr/bin/xcodebuild`
npmPackages:
      `react: 16.4.1 => 16.4.1 `
      `react-native: 0.56.0 => 0.56.0 `

**Comment by mattijsf** [comment]
I got error as well after updating to Xcode 10. Build goes fine for simulators but not for Generic / Physical Devices. This is blocking development using Xcode 10 atm :( Which wouldn't be a problem if we could use Xcode 9.4. But that means we cannot develop with iOS 12.

```
Environment:
  OS: macOS High Sierra 10.13.6
  Node: 8.3.0
  Yarn: 1.3.2
  npm: 6.4.1
  Watchman: 4.9.0
  Xcode: Xcode 10.0 Build version 10A255
  Android Studio: 3.1 AI-173.4907809

Packages: (wanted => installed)
  react: ^16.3.1 => 16.3.1
  react-native: ^0.55.4 => 0.55.4
```

**Comment by kelset** [comment]
@mattijsf have you tried with 0.57? As you can see on #19573, we are not fully XCode 10 ready, and any help in fixing the remaining issues is surely welcome.

**Comment by mattijsf** [comment]
I haven't. I can try but I don't expect this will make a difference. I think this has something to do with the glog library:

During the build / configure step of that library it outputs the follow:

`cd node_modules/react-native/third-party/glog-0.3.4/ && ./configure`

```bash
...
checking for sys/ucontext.h... yes
checking how to access the program counter from a struct ucontext... uc_mcontext->__ss.__rip
checking that generated files are newer than configure... done
...
```

Looking at the relevant script it seems that they do some one-off trial 'n error for obtaining this "program counter". Once it finds a working field it writes it to `node_modules/react-native/third-party/glog-0.3.4/src/config.h`
> ...
> #define PC_FROM_UCONTEXT uc_mcontext->__ss._r
> ...

Which in turn only works for simulators (which is the same architecture as macOS: darwin/x86) but no longer for devices (darwin/arm)

**Comment by mattijsf** [comment]
I found a workaround but its untested:
It involves manually editing `react-native/node_modules/react-native/third-party/glog-0.3.4/src/config.h`:
Replace:
```c
/* How to access the PC from a struct ucontext */
#define PC_FROM_UCONTEXT uc_mcontext->__ss.__rip
```
With:
```c
/* How to access the PC from a struct ucontext */
#if defined(__arm__) || defined(__arm64__)
#define PC_FROM_UCONTEXT uc_mcontext->__ss.__pc
#else
#define PC_FROM_UCONTEXT uc_mcontext->__ss.__rip
#endif
```

Using this it will compile for both targets. My knowledge is pretty much absent regarding the use of these values. Except that `__darwin_arm_thread_state64` does provide the information that this pointer is in fact the "program counter":
`/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/mach/arm/_structs.h` @ line 140:
```
_STRUCT_ARM_THREAD_STATE64
{
	__uint64_t    __x[29];	/* General purpose registers x0-x28 */
	__uint64_t    __fp;		/* Frame pointer x29 */
	__uint64_t    __lr;		/* Link register x30 */
	__uint64_t    __sp;		/* Stack pointer x31 */
	__uint64_t    __pc;		/* Program counter */
	__uint32_t    __cpsr;	/* Current program status register */
	__uint32_t    __pad;    /* Same size for 32-bit or 64-bit clients */
};
```

**Comment by kelset** [comment]
Well in 0.57 we use [glog 0.3.5](https://github.com/facebook/react-native/blob/3b688ae6b294e94f11e8eff425adf698e5b3783c/third-party-podspecs/glog.podspec), maybe that script is fixed there?

**Comment by mattijsf** [comment]
I updated to 0.57.0 and did a full clean using:

```bash
rm -rf node_modules
npm cache clear --force ; watchman watch-del-all ; rm -rf $TMPDIR/react-* ; rm -rf ~/.rncache
```

But the error remains the same. The above workaround still seems to work though.

**Comment by mattijsf** [comment]
The problems seems to be resolved now. I'm not entirely sure but I think it had to do with the way I was installing `glog`.

With earlier iterations (both Xcode / RN) I was running the glog `./configure` script manually (due to Xcode failing to do it properly) like so:
https://github.com/facebook/react-native/issues/14382#issuecomment-420190830

That actually produces a broken `config.h` because the script:
https://github.com/facebook/react-native/blob/v0.57.0/scripts/ios-configure-glog.sh#L44-L51
Was actually building glog with `./configure --host arm-apple-darwin` and taking into account patching the `PC_FROM_UCONTEXT` macro.

I learned that Xcode failed building / installing the third-party dependencies only at the first build and that subsequent builds actually takes care of installing `glog` correctly: https://github.com/facebook/react-native/issues/21168#issuecomment-422502787

**Comment by Monte9** [comment]
Running into this issue too. Just updated to Xcode 10 to test on an iOS 12 device with React Native `0.54.4`.

We currently don't have any plans to update to the latest react native version.. so I was wondering if there is any other fix for it?

**Comment by MayeuldP** [comment]
@Monte9 
https://github.com/facebook/react-native/issues/21168#issuecomment-423103382 
Following this one, and doing a `Product clean again after through Xcode` solved the problem.

**Comment by Monte9** [comment]
Ok, great @MayeuldP! Thanks for the suggestion. 

For now, I ended up downgrading to Xcode 9 & iOS 11 just cause I can't afford to spend time debugging the issue. In case anyone else is looking to do so, I found [this helpful guide](https://www.macworld.co.uk/how-to/apple/downgrade-ios-12-3664217/)

**Comment by yllow** [comment]
@Monte9 have you tried Legacy Build System? worked for me

**Comment by Monte9** [comment]
I did. But that still thew a few errors, so I didn't debug them further.

**Comment by hramos** [comment]
The original issue appears to be resolved in 0.57. For anything else related to third party dependencies, the issue you're looking for is #20774.
